libs/libc/sched/task_startup.c

```diff
@@ -98,6 +98,11 @@ static void cxx_initialize(void)
 
   if (inited == 0)
     {
+#if defined(CONFIG_ARCH_SIM) && defined(CONFIG_HOST_MACOS)
+      extern void macho_call_saved_init_funcs(void);
+
+      macho_call_saved_init_funcs();
```

Comment:
can we construct _sinit/_einit section in arch/sim manually to avoid the macos hack in the common code?

Reply:
the addresses of _sinit/_einit are link-time constants. i can't think of a reasonable way to construct them on runtime.

Reply:
How about we extract the initialization to a weak function? so we can provide a new implemenation under arch/sim.

Reply:
i'm not sure if it's a good idea to use weak functions.
i pushed a version with usual Kconfig as it seemed enough to me.

Reply:
> the addresses of _sinit/_einit are link-time constants. i can't think of a reasonable way to construct them on runtime.

We can put macho_call_saved_init_funcs into this section, so the callchain is:
nxtask_startup->cxx_initialize->macho_call_saved_init_funcs->each constructor

Reply:
maybe.
but we still need to provide _sinit and _einit with the right addresses. maybe using a linker script or something.
i don't feel it's worth the complexity.

Reply:
We can use alias here to avoid the link script.

Reply:
alias? can you explain?

anyway, emulating sinit/einit thing does not seem particularly attractive to me.
there are multiple platform/object format dependent ways to represent these stuff.
sinit/einit is just one of them.
i suspect we will need yet another version for windows.

Reply:
https://gcc.gnu.org/onlinedocs/gcc/Common-Variable-Attributes.html. Yes, I agree it's better to replace cxx_initilize directly.